core/src/main/java/com/linecorp/armeria/server/http/auth/OAuth1aAuthService.java

```diff
@@ -0,0 +1,84 @@
+package com.linecorp.armeria.server.http.auth;
+
+import java.util.Map;
+import java.util.function.Function;
+import java.util.function.Predicate;
+import java.util.regex.Matcher;
+import java.util.regex.Pattern;
+
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+
+import com.google.common.base.Strings;
+import com.google.common.collect.Maps;
+
+import com.linecorp.armeria.common.http.HttpHeaderNames;
+import com.linecorp.armeria.common.http.HttpHeaders;
+import com.linecorp.armeria.common.http.HttpRequest;
+import com.linecorp.armeria.common.http.HttpResponse;
+import com.linecorp.armeria.server.Service;
+
+/**
+ * An {@link HttpAuthService} with <a href="https://oauth.net/core/1.0a/">Oauth1a authorization</a> support.
+ * */
+public class OAuth1aAuthService extends HttpAuthService<OAuth1aToken> {
+
+    private static final Logger logger = LoggerFactory.getLogger(OAuth1aAuthService.class);
+
+    private static final Pattern AUTHORIZATION_HEADER_PATTERN = Pattern.compile(
+            "(?i)oauth\\s+(?<parameters>\\S+)");
+
+    /**
+     * Creates a new HTTP basic authorization {@link Service} decorator using the specified {@code predicate}.
+     */
+    public static Function<Service<? super HttpRequest, ? extends HttpResponse>,
+            OAuth1aAuthService> newDecorator(Predicate<? super OAuth1aToken> predicate) {
+        return service -> new OAuth1aAuthService(service, predicate);
+    }
+
+    OAuth1aAuthService(Service<? super HttpRequest, ? extends HttpResponse> delegate,
+                       Predicate<? super OAuth1aToken> predicate) {
+        super(delegate, predicate);
+    }
+
+    @Override
+    public String toString() {
+        StringBuilder sb = new StringBuilder("OAuth1aAuthService(predicate:");
```

Comment:
How about using `MoreObjects.toStringHelper()`? Also, I'm not sure we should use `hashCode()` here; a predicate implementation might have a proper `toString()` implementation.